File: plugins/logger/logger.js
/*
 * logger.js -- plugin to log everything
 *
 * Copyright (c) 2013-2022 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

// Plugin information.
info = {
	author: "David Demelier <markand@malikania.fr>",
	license: "ISC",
	summary: "A plugin to log everything",
	version: "@irccd_VERSION@"
};

// Modules.
var Directory = Irccd.Directory;
var File = Irccd.File;
var Logger = Irccd.Logger;
var Plugin = Irccd.Plugin;
var Util = Irccd.Util;

/**
 * All available templates.
 */
Plugin.templates = {
	"join":         "%H:%M:%S >> #{nickname} joined #{channel}",
	"kick":         "%H:%M:%S :: #{target} has been kicked by #{nickname} [reason: #{reason}]",
	"me":           "%H:%M:%S * #{nickname} #{message}",
	"message":      "%H:%M:%S #{nickname}: #{message}",
	"invite":       "%H:%M:%S #{nickname} invited you on #{channel}",
	"mode":         "%H:%M:%S :: #{nickname} set mode #{channel} #{mode} #{args}",
	"notice":       "%H:%M:%S [notice] #{channel} (#{nickname}) #{message}",
	"part":         "%H:%M:%S << #{nickname} left #{channel} [#{reason}]",
	"query":        "%H:%M:%S #{nickname}: #{message}",
	"topic":        "%H:%M:%S :: #{nickname} changed the topic of #{channel} to: #{topic}"
};

function keywords(server, channel, origin, extra)
{
	var kw = {
		"server": server.toString(),
		"channel": channel,
		"origin": origin,
		"nickname": Util.splituser(origin)
	};

	for (var key in extra)
		kw[key] = extra[key];

	return kw;
}

function write(fmt, args)
{
	var path = Util.format(Plugin.config["file"], args);
	var directory = File.dirname(path);

	// Try to create the directory.
	if (!File.exists(directory)) {
		Logger.debug("creating directory: " + directory);
		Directory.mkdir(directory);
	}

	Logger.debug("opening: " + path);

	var str = Util.format(Plugin.templates[fmt], args);
	var file = new File(path, "a");

	file.write(str + "\n");
}

function onLoad()
{
	if (Plugin.config["file"] === undefined)
		throw new Error("Missing 'file' option");
}

function onInvite(server, origin, channel)
{
	origin = origin.toLowerCase();
	channel = channel.toLowerCase();

	write("invite", keywords(server, channel, origin));
}

function onJoin(server, origin, channel)
{
	origin = origin.toLowerCase();
	channel = channel.toLowerCase();

	write("join", keywords(server, channel, origin));
}

function onKick(server, origin, channel, target, reason)
{
	origin = origin.toLowerCase();
	channel = channel.toLowerCase();

	write("kick", keywords(server, channel, origin, {
		"target":       target,
		"reason":       reason
	}));
}

function onMe(server, origin, channel, message)
{
	origin = origin.toLowerCase();
	channel = channel.toLowerCase();

	write("me", keywords(server, channel, origin, {
		"message": message,
	}));
}

function onMessage(server, origin, channel, message)
{
	origin = origin.toLowerCase();
	channel = channel.toLowerCase();

	write("message", keywords(server, channel, origin, {
		"message": message,
	}));
}

function onMode(server, origin, channel, mode, args)
{
	origin = origin.toLowerCase();
	channel = channel.toLowerCase();

	write("mode", keywords(server, channel, origin, {
		"mode":         mode,
		"args":         args.join(" "),
	}));
}

function onNick(server, origin, nickname)
{
	// TODO: write for all servers/channels a log entry
}

function onNotice(server, origin, channel, notice)
{
	origin = origin.toLowerCase();
	channel = channel.toLowerCase();

	write("notice", keywords(server, channel, origin, {
		"message": notice,
	}));
}

function onPart(server, origin, channel, reason)
{
	origin = origin.toLowerCase();
	channel = channel.toLowerCase();

	write("part", keywords(server, channel, origin, {
		"reason": reason,
	}));
}

function onTopic(server, origin, channel, topic)
{
	origin = origin.toLowerCase();
	channel = channel.toLowerCase();

	write("topic", keywords(server, channel, origin, {
		"topic": topic
	}));
}

File: libirccd-daemon/irccd/daemon/logger.cpp
/*
 * logger.cpp -- irccd logging
 *
 * Copyright (c) 2013-2020 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <irccd/sysconfig.hpp>

#include <cassert>
#include <fstream>
#include <iostream>
#include <streambuf>

#include "logger.hpp"

#if defined(IRCCD_HAVE_SYSLOG)
#	include <syslog.h>
#endif // !IRCCD_HAVE_SYSLOG

namespace irccd::daemon::logger {

void logger::debug(const std::string& line)
{
	// Print only in debug mode, the buffer is flushed anyway.
#if !defined(NDEBUG)
	parent_.write_debug(parent_.filter_->pre_debug(category_, component_, line));
#else
	(void)line;
#endif
}

void logger::info(const std::string& line)
{
	// Print only if verbose, the buffer will be flushed anyway.
	if (parent_.verbose_)
		parent_.write_info(parent_.filter_->pre_info(category_, component_, line));
}

void logger::warning(const std::string& line)
{
	parent_.write_warning(parent_.filter_->pre_warning(category_, component_, line));
}

logger::logger(sink& parent, level level, std::string_view category, std::string_view component) noexcept
	: std::ostream(this)
	, level_(level)
	, parent_(parent)
	, category_(category)
	, component_(component)
{
	assert(level >= level::debug && level <= level::warning);
}

int logger::sync()
{
	std::string buffer = str();
	std::string::size_type pos;

	while ((pos = buffer.find("\n")) != std::string::npos) {
		auto line = buffer.substr(0, pos);

		// Remove this line.
		buffer.erase(buffer.begin(), buffer.begin() + pos + 1);

		switch (level_) {
		case level::debug:
			debug(line);
			break;
		case level::info:
			info(line);
			break;
		case level::warning:
			warning(line);
			break;
		default:
			break;
		}
	}

	str(buffer);

	return 0;
}

void console_sink::write_info(const std::string& line)
{
	std::cout << line << std::endl;
}

void console_sink::write_warning(const std::string& line)
{
	std::cerr << line << std::endl;
}

void console_sink::write_debug(const std::string& line)
{
	std::cout << line << std::endl;
}

file_sink::file_sink(std::string normal, std::string errors)
	: output_normal_(std::move(normal))
	, output_error_(std::move(errors))
{
}

void file_sink::write_info(const std::string& line)
{
	std::ofstream(output_normal_, std::ofstream::out | std::ofstream::app) << line << std::endl;
}

void file_sink::write_warning(const std::string& line)
{
	std::ofstream(output_error_, std::ofstream::out | std::ofstream::app) << line << std::endl;
}

void file_sink::write_debug(const std::string& line)
{
	std::ofstream(output_normal_, std::ofstream::out | std::ofstream::app) << line << std::endl;
}

void silent_sink::write_info(const std::string&)
{
}

void silent_sink::write_warning(const std::string&)
{
}

void silent_sink::write_debug(const std::string&)
{
}

#if defined(IRCCD_HAVE_SYSLOG)

syslog_sink::syslog_sink()
{
	openlog("irccd", LOG_PID, LOG_DAEMON);
}

syslog_sink::~syslog_sink()
{
	closelog();
}

void syslog_sink::write_info(const std::string& line)
{
	syslog(LOG_INFO | LOG_USER, "%s", line.c_str());
}

void syslog_sink::write_warning(const std::string& line)
{
	syslog(LOG_WARNING | LOG_USER, "%s", line.c_str());
}

void syslog_sink::write_debug(const std::string& line)
{
	syslog(LOG_DEBUG | LOG_USER, "%s", line.c_str());
}

#endif // !IRCCD_HAVE_SYSLOG

sink::sink()
	: filter_(new filter)
{
}

auto sink::is_verbose() const noexcept -> bool
{
	return verbose_;
}

void sink::set_verbose(bool mode) noexcept
{
	verbose_ = mode;
}

void sink::set_filter(filter& filter) noexcept
{
	filter_ = &filter;
}

auto sink::info(std::string_view category, std::string_view component) -> logger
{
	return logger(*this, logger::level::info, category, component);;
}

auto sink::warning(std::string_view category, std::string_view component) -> logger
{
	return logger(*this, logger::level::warning, category, component);;
}

auto sink::debug(std::string_view category, std::string_view component) -> logger
{
	return logger(*this, logger::level::debug, category, component);;
}

auto filter::pre(std::string_view category,
                 std::string_view component,
                 std::string_view message) const -> std::string
{
	std::ostringstream oss;

	oss << category;

	if (!component.empty())
		oss << " " << component;

	oss << ": ";
	oss << message;

	return oss.str();
}

auto filter::pre_debug(std::string_view category,
                       std::string_view component,
                       std::string_view message) const -> std::string
{
	return pre(category, component, message);
}

auto filter::pre_info(std::string_view category,
                      std::string_view component,
                      std::string_view message) const -> std::string
{
	return pre(category, component, message);
}

auto filter::pre_warning(std::string_view category,
                         std::string_view component,
                         std::string_view message) const -> std::string
{
	return pre(category, component, message);
}

} // !irccd::daemon::logger
